init/tmpfiles.py

```diff
@@ -0,0 +1,606 @@
+            except:
+                print(f'failed to chmod {path!r}')
+
+
+class GlobAction(Action):
+    """Action that takes a glob rather than a path."""
+    def apply(self, pattern):
+        for path in glob.glob(pattern):
+            self.apply_one(path)
+
+
+class FileCreate(Action):
+    """f - create file with optional content"""
+    def apply_one(self, path):
+        if not os.path.lexists(path):
+            with open(path, 'w') as f:
+                f.write(self.arg if self.arg is not None else '')
+        self._chown_and_chmod(path)
+
+
+class FileCreateOrTruncate(Action):
+    """f+ - create or truncate file, with optional content"""
+    def apply_one(self, path):
+        with open(path, 'w') as f:
+            f.write(self.arg if self.arg is not None else '')
+        self._chown_and_chmod(path)
+
+
+class FileWrite(GlobAction):
+    """w - write to file"""
+    # TODO interpret C-style blackslashes in argument.  Also for
+    # other actions (f, f+, w+, ...)
+    def apply_one(self, path):
+        with open(path, 'w') as f:
+            f.write(self.arg if self.arg is not None else '')
+        self._chown_and_chmod(path)
+
+
+class FileAppend(GlobAction):
+    """w+ - append to file"""
+    def apply_one(self, path):
+        with open(path, 'a') as f:
+            f.write(self.arg if self.arg is not None else '')
+        self._chown_and_chmod(path)
+
+
+class DirCreateAndCleanup(Action):
+    """d - create and cleanup directory"""
+    def apply_one(self, path):
+        if not os.path.lexists(path):
+            os.makedirs(path)
+            self._chown_and_chmod(path)
+
+
+class DirCreateAndRemove(DirCreateAndCleanup):
+    """D - create and remove directory"""
+    # No additional behaviour, until --remove gets implemented
+
+
+class DirCleanup(GlobAction):
+    """e - create and remove directory"""
+    def apply_one(self, path):
+        if os.path.isdir(path):
+            self._chown_and_chmod(path)
+
+
+class SubvolumeCreate_v(DirCreateAndCleanup):
+    """v - create subvolume or directory"""
+    # ignore subvolume behaviour
+
+
+class SubvolumeCreate_q(DirCreateAndCleanup):
+    """q - create subvolume or directory"""
+    # ignore subvolume behaviour
+
+
+class SubvolumeCreate_Q(DirCreateAndCleanup):
+    """Q - create subvolume or directory"""
+    # ignore subvolume behaviour
+
+
+# TODO p, p+
+
+
+class SymlinkCreate(Action):
+    """L - create symlink"""
+    def apply_one(self, path):
+        if not os.path.exists(path):
+            if self.arg is None:
+                # TODO link to /usr/share/factory/FILE
+                # (see tmpfiles.d(5) for details)
+                raise RuntimeError("symlink target not specified")
+            os.symlink(self.arg, path)
+
+
+class SymlinkRecreate(SymlinkCreate):
+    """L+ - [re]create symlink"""
+    def apply_one(self, path):
+        if os.path.exists(path):
+            if not os.path.islink(path) and os.path.isdir(path):
```

Comment:
This will call ``stat()`` three times. Is this a performance problem?

Reply:
I don't think so :)  I prefer the readability of ^^^, unless there are measurements to show that it's a problem.